### PR TITLE
Add comments about Motion deadlock hazards for Joins

### PR DESCRIPTION
**Pull request by hardikar**
This comes from a discussion on the GPDB mailing list about need of
the prefetch_inner parameter in Joins and the deadlock case it is
trying to prevent.

### CODE REVIEW

File: src/backend/cdb/cdbllize.c [hunk 1]
```diff
@@ -1366,7 +1366,37 @@ motion_sanity_walker(Node *node, sanity_result_t *result)
...
+	 *
+	 * A cannot ACK C because it is waiting to finish retrieving inner tuples
+	 * from D. B cannot ACK D because it is waiting for it's first outer tuple
+	 * from C before accepting any inner tuples. This forms a circular
+	 * dependency resulting in a deadlock : C -> A -> D -> B -> C.
+	 *
+	 * We avoid this by pre-fetching all the inner tuples in such cases and
+	 * materializing them in some fashion, before moving on to outer_tuples.
+	 * This effectively breaks the cycle and prevents deadlock.
+	 */
```
Comment (asimrp) [inline src/backend/cdb/cdbllize.c]: Can we please add a pointer to the mailing list discussion? https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/gMa1tW0x_fk/wuzvGXBaBAAJ

**Comment by hlinnaka** [comment]
+1.

Not sure this is the best place for the comment, as motion_sanity_walker() is just an additional checker for the issue. Will people find it from there, when wondering what prefetch_inner or cdb_strict is? I'm not sure what a better place would be, though.

There are several places where we deal with this issue:
* in nodeMotion.c, the checks on active_recv_id are related to this (search for "DEADLOCK HAZARD")
* in nodeNestLoop.c, nodeHashjoin.c, nodeMergejoin.c, the prefetch_inner-related code, with the comments referring MPP-3300.
* in nodeMaterial.c, code that checks cdb_strict
* in createplan.c, checks on motionHazard to set cdb_strict

It would be nice to add a reference in all those places, to the longer comment you're adding that explains the whole issue.

**Comment by hsyuan** [comment]
Should we add the `deadlock` test into schedule file? otherwise, it is not going to be enabled in ICG test.

File: src/test/regress/sql/deadlock.sql [hunk 2]
```diff
@@ -0,0 +1,42 @@
+-- 
+-- Test queries that can lead to deadlock. This is in a different file to
+-- isolate the test & quickly identify it if ICG doesn't complete for a very
+-- long time.
+--
+
+
+create schema deadlock;
+
```
Comment (hsyuan) [inline src/test/regress/sql/deadlock.sql]: Add `set search_path = deadlock; ` here. otherwise, it is going to use the default schema.

**Comment by hardikar** [comment]
@hlinnaka I see what you mean - I wasn't sure of the best place either. What I can say is that I definitely looked at `motion_sanity_walker()` when I was trying to figure out how a deadlock can occur - so I suppose its a reasonable expectation others will too. 

Nevertheless - Adding a reference is even better. Added it to all the locations you mentioned.

**Pull request merged**